File: 2022-2023/sourcecode/lab/lab11/VeryNobleAppJava/app/src/main/java/pl/udu/uwr/pum/verynobleappjava/api/RetrofitInstance.java
package pl.udu.uwr.pum.verynobleappjava.api;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitInstance {
    private RetrofitInstance(){}

    private static volatile NobelPrizeApi api;
    private static final String baseUrl = "https://api.nobelprize.org/";

    public static NobelPrizeApi getApi() {
        if (api == null) {
            synchronized (RetrofitInstance.class) {
                if (api == null) {
                    HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
                    interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
                    OkHttpClient client = new OkHttpClient.Builder()
                            .addInterceptor(interceptor)
                            .build();
                    api = new Retrofit.Builder()
                            .baseUrl(baseUrl)
                            .addConverterFactory(GsonConverterFactory.create())
                            .client(client)
                            .build().create(NobelPrizeApi.class);
                }
            }
        }
        return api;
    }
}

File: 2022-2023/sourcecode/lab/lab11/VeryNobleAppKotlin/app/src/main/java/pl/udu/uwr/pum/verynobleappkotlin/ui/api/RetrofitInstance.kt
package pl.udu.uwr.pum.verynobleappkotlin.ui.api

import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import pl.udu.uwr.pum.verynobleappkotlin.ui.util.baseUrl
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitInstance {
    val api: NobelPrizeApi by lazy {
        val logging = HttpLoggingInterceptor()
        logging.setLevel(HttpLoggingInterceptor.Level.BODY)
        val client = OkHttpClient.Builder()
            .addInterceptor(logging)
            .build()
        Retrofit.Builder()
            .baseUrl(baseUrl)
            .addConverterFactory(GsonConverterFactory.create())
            .client(client)
            .build()
            .create(NobelPrizeApi::class.java)
    }
}